### DISCUSSION TITLE
Interceptors no NestJS com Class-validator

### DISCUSSION THREAD

**Discussion by izaque** [balta-io/forum#309] [Tira Dúvidas] [answered]
No curso 7180, aula de interceptors, como criar algo semelhante para funcionar com class-validator e DTOs?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @izaque, como vai?

Qual seria o contexto? O que precisa validar? Consegue detalhar com algum código?

**Reply by izaque** [reply]
> Bom dia, @izaque, como vai?
> 
> Qual seria o contexto? O que precisa validar? Consegue detalhar com algum código?

Balta, boa tarde. Eu creio que consegui fazer funcionar. Eu precisei adicionar um contructor do tipo Any para o DTO. Eu só não consegui registrá-lo no app.module.ts ou no main.ts, mas de qualquer forma está funcionando em todos os métodos dos controllers em que faço a chamada. Segue o código:

// validator.interceptor.ts

import { Injectable, NestInterceptor, ExecutionContext, CallHandler, BadRequestException, HttpException, HttpStatus } from '@nestjs/common';
import { Observable } from 'rxjs';
import { plainToClass } from 'class-transformer';
import { validate } from 'class-validator';
import { Result } from 'src/result.model';


@Injectable()
export class ValidatorInterceptor implements NestInterceptor {
    constructor(private readonly targetDto: any) { }

    async intercept(context: ExecutionContext, next: CallHandler): Promise<Observable<any>> {
        const request = context.switchToHttp().getRequest();
        const body = request.body;

        const dtoInstance = plainToClass(this.targetDto, body);

        const errors = await validate(dtoInstance);

        const qtdeErros = errors.map(error => Object.values(error.constraints)).flat().length;
        // console.log('Erros: ', errors);

        if (errors.length > 0) {
            const errorMessages = errors.map(error => Object.values(error.constraints)).flat();
            throw new HttpException(new Result('Entrada de dados inválida. Por favor, verifique os parâmetros necessários para a requisição e detalhes do erro abaixo.', false, null, errorMessages), HttpStatus.BAD_REQUEST);
            // throw new BadRequestException('Os seguintes erros de validação foram encontrados: ' + errorMessages);
        }

        return next.handle();
    }
}

// create-unidade.dto.ts

import { ApiProperty } from "@nestjs/swagger";

import {
    IsNotEmpty,
    IsString,
    Length,
} from 'class-validator';

export class CreateUnidadeDto {

    @IsString()
    @IsNotEmpty({ message: "O campo Nome deve ser informado." })
    @Length(2, 50, { message: "O Nome deve estar entre 3 e 50 caracteres." })
    @ApiProperty({
        description: 'Nome da Unidade',
        minimum: 3,
        maximum: 50,
        example: "Quilograma"
    })
    nome: string;

    @IsString()
    @IsNotEmpty({ message: "O campo Sigla deve ser informado." })
    @Length(2, 3, { message: "O Nome deve estar entre 2 e 3 caracteres." })
    @ApiProperty({
        description: 'Sigla da Unidade',
        minimum: 2,
        maximum: 3,
        example: "KG"
    })
    sigla: string;
}

// Controller:
// unidades.controller.ts

import { Body, CacheInterceptor, Controller, Delete, Get, HttpException, HttpStatus, Param, ParseIntPipe, Patch, Post, UseInterceptors } from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { Result } from 'src/result.model';
import { ValidatorInterceptor } from 'src/shared/interceptors/validator.interceptor';
import { CreateUnidadeDto } from './dto/create-unidade.dto';
import { UpdateUnidadeDto } from './dto/update-unidade.dto';
import { UnidadesService } from './unidades.service';

@Controller('unidades')
@ApiTags('Unidades')
export class UnidadesController {
  constructor(private readonly unidadesService: UnidadesService) { }

  @Post()
  @UseInterceptors(new ValidatorInterceptor(CreateUnidadeDto)) // Aplica o interceptor à rota de criação de Unidades
  async create(@Body() createUnidadeDto: CreateUnidadeDto): Promise<CreateUnidadeDto> {
    try {
      return await this.unidadesService.create(createUnidadeDto);
    } catch (error) {
      throw new HttpException(new Result(null, false, null, error), HttpStatus.BAD_REQUEST);
    }
  }

}

**Reply by izaque** [reply]
Aprendendo TypeScript/Javascript junto com o NestJs. Caso tenha alguma consideração para a melhora do código pode me informar. 
Obrigado.

**Reply by andrebaltieri** [reply]
Wow, sensacional... parabéns pelo empenho 🚀🚀🚀🚀
